src/main/java/frc/robot/subsystems/swerve/PhoenixOdometryThread.java

```diff
@@ -51,22 +53,40 @@ public static PhoenixOdometryThread getInstance() {
   private PhoenixOdometryThread() {
     setName("PhoenixOdometryThread");
     setDaemon(true);
-    start();
+  }
+
+  @Override
+  public void start() {
+    if (timestampQueues.size() > 0) {
+      super.start();
+    }
   }
 
   public Queue<Double> registerSignal(ParentDevice device, StatusSignal<Double> signal) {
-    isCANFD = CANBus.isNetworkFD(device.getNetwork());
-    Queue<Double> queue = new ArrayBlockingQueue<>(100);
+    Queue<Double> queue = new ArrayDeque<>(100);
```

Comment:
This seems like a potentially dangerous change in behavior. The `ArrayBlockingQueue` is size limited and blocks the thread when a 101st item is added. This will never block and is unbounded. I assume this got swapped out because the thread was blocking when we were pulling items from the queue too slowly? That speaks to a problem that should probably be solved (does the max queue size need to be bigger? Do we need to drop samples? Do we need to process faster?) vs being worked around.

Reply:
so this is from 6328's template code, which is what we've been using for the multithread stuff. it seems like they addressed that today [here](https://github.com/Mechanical-Advantage/AdvantageKit/issues/73) and [here.](https://github.com/Mechanical-Advantage/AdvantageKit/commit/a6fc30c01ee1ab523ca06da96308a34d2b3bc5a2) I can use that if you think it looks alright, but I'm not super clear on how/why this works so would you be available to have a more detailed discussion on this sometime?

Reply:
Yeah, looks like they just reverted this change. Both a smaller queue and back to blocking.

Happy to talk in person or slack, but will try to give a summary here.

Queues have a few options when you try to put things in them. They can accept as much as you give or have a limit. When they hit a limit they can drop things on the floor, refuse to take them, or wait until they've got room and then add. `offer` is a nonblocking method by definition and returns `false` if it can't fit the item so we're giving the implementation an option to reject too-much-data.

So lets talk options we've seen.

 [ArrayDeque](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/ArrayDeque.html) is unbounded ("no capacity restrictions; they grow as necessary) so `offer` will by definition never return `false`. That number you pass to the constructor is the initial capacity (so it allocates memory for 100 items all at once). This class is explicitly not thread safe, so you need to roll your own synchronization.

In contrast, [BlockingArrayQueue](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/concurrent/ArrayBlockingQueue.html) is defined with a *limit*. `offer` will return false if we try to go over the limit and it also has a `put` method (part of the [BlockingQueue](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/concurrent/BlockingQueue.html) interface) which will pause the thread until there's room in the queue. That would be weird in a single-threaded environment but makes lots of sense if it's an extra worker thread. In any case, we're never going to block in this use because offer is being called. Effectively we've opted into limiting the size and dropping the data on the floor if we overrun. Honestly, I'm not sure that's the right choice either - probably we want the *most recent* data, not the first into the queue if we overrun. Probably a better choice is a circular buffer / ring buffer like the [EvictingQueue](https://guava.dev/releases/23.0/api/docs/com/google/common/collect/EvictingQueue.html) implemented in Java which will drop the oldest data rather than the newest when it fills.